### ISSUE TITLE
Redesign/implement class/modules that are created in gRPC PoC 

### ISSUE THREAD

**Issue by srcansiz** [fedbiomed/fedbiomed#856]
- [x] implement gRPC communications
- [x] refactor new gRPC classes
- refactor modified and impacted classes
     - [x] researcher: Job, Experiment, Requests
     - [x] node: Node, etc.
     - [x] common
     - [x] scripts and envs for dev mode
     - [x] gui ?

- [x] adapt for merging with dependencies import patch (#870)
  - choosed solution: move `configure_dependencies()` from `post_init()` to the end of `init()`
- [x] partial refactor of `Job()` and `Experiment()` for accepting training plan parameter only as a `TrainingPlan` class (not anymore a path to a file or an object/instance of a class)

- [x] cleaning: clean out any MQTT/restful unused code (+ comments, docstrings)

- [x] logger: adapt logger
    * adapt log API for making node => researcher optional (by default: node local logging)
    * extend logger to send log messages to one specific researcher_id `logger.error(researcher_id=1234)`. By default, only local logging.
- ~~[ ] message send queues on node & researcher:~~ => not implemented for now, create separate issue ?
   * ~~choose and implement robustness policy, eg: keep using `queue.put_nowait()` if node is in CONNECTED state ; introduce a MAX_SEND_DELAY timer after `queue.get()` so that all messages older than delay are discarded ; retry sending message if sending error (`queue.task_done()` after sending successful) with a retry delay~~

- robustness: test & handle scenarios of failure/interrupt in node or researcher (while setting up communications, searching, training, normal termination ; notebook + python script)
    - [x] see crash scenario in comment below
    - [x] node does not terminate after receiving an unknown gRPC error (see comment below)

- ~~[ ] encrypt communications (without parties authentication, but prevents passive MITM attacks == honest but curious)~~ => in #875
- ~~[ ] test with all notebooks (use new CI end2end tests ?)~~ => in #857
- ~~[ ] unit tests: adapt & add~~ => in #857
- ~~[ ] update documentation (tutorials, user guide, security, matrix, etc.)~~ => in #859
- ~~[ ] update and test VPN/container mode~~ => in #858

**Comment by mvesin** [comment]
Other robustness issue: when receiving an unknown gRPC error, this should cause the node to terminate and exit. This is not the case

```
...
2023-10-02 18:25:28,318 fedbiomed ERROR - Unexpected error raised by researcher gRPC server. This is probably due to bug on the researcher side: <AioRpcError of RPC that terminated with:
	status = StatusCode.UNKNOWN
	details = "Unexpected <class 'AttributeError'>: 'Requests' object has no attribute '_monitor_message_callback'"
	debug_error_string = "UNKNOWN:Error received from peer  {grpc_message:"Unexpected <class \'AttributeError\'>: \'Requests\' object has no attribute \'_monitor_message_callback\'", grpc_status:2, created_time:"2023-10-02T18:25:28.318137775+02:00"}"
>
2023-10-02 18:25:28,318 fedbiomed CRITICAL - An exception raised by running tasks within GrpcClients. This will stop gRPC client. The exception: FedbiomedCommunicationError. Error message: ErrorNumbers.FB628: Sender stopped due to error on the researcher side
2023-10-02 18:25:28,318 fedbiomed INFO - Node is stopped!

(remains blocked until ctrl-C - or timeout ?)
```

**Comment by mvesin** [comment]
> Other robustness issue: when receiving an unknown gRPC error, this should cause the node to terminate and exit. This is not the case

After discussion with @srcansiz we robustified twofold:
- when receiving a gRPC error in a node side task for researcher (`Sender` or `TaskListener`) : no need to raise exception and exit, just relaunch the task (after waiting, for task listener case)
- when receiving another type of error in a node side task for researcher (`Sender` or `TaskListener`) : keep raising an error. Then both tasks exit. We added an optional callback for an action at that time. Use it to trigger a SIGTERM to the current process, so that the node exits when all researchers are failed (instead of keeping running, waiting on queue, and user has to ctrl-C to exit the node)
